### PR TITLE
Add Railgun projectile

### PR DESCRIPTION
**Pull request by forcecore**
I hope this looks good
https://youtu.be/tHJgJtj-Utk

### CODE REVIEW

**Comment by rob-v** [comment]
It looks great. I played one game with it, works fine with units, doesn't work with buildings. You need to force attack behind building to do damage:
obrazok

**Comment by pchote** [comment]
> doesn't work with buildings. You need to force attack behind building to do damage:

That is caused by the ongoing targeting rework, and not part of this PR.

File: OpenRA.Mods.Common/Graphics/RailgunRenderable.cs
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/RailgunRenderable.cs]: I haven't tested this PR, but I expect that rendering individual line segments with alpha will cause artefacts at the joins where they overlap.  Pull this outside the loop and use the `IEnumerable<float3>` overload to have the renderer properly join up the segments.

**Comment by forcecore** [comment]
Now that TS mod works with new bleed head, rebased. Helix line segment drawing pulled out of the loop.

**Comment by reaperrr** [comment]
> It looks great. I played one game with it, works fine with units, doesn't work with buildings. You need to force attack behind building to do damage:

Will be fixed once #13464 is finished and merged.

File: OpenRA.Mods.Common/Graphics/RailgunRenderable.cs
Comment (GraionDilach) [inline OpenRA.Mods.Common/Graphics/RailgunRenderable.cs]: This still mentions the rad beam.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (GraionDilach) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: This needs a rewording, but I have noidea to what to.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: We could mention that this will overwrite `BeamColor`.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: Similar case here.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: `// Fortunately, railguns shoot projectiles which don't track the target.`

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: `Thickness`

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: This seems to be the same description as the one of `HelixPitch`?

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: You should omit those braces around a single-line statement.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: Why do we need a separate method here?

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: Can we please change `Cnt` to `Count`? `Cnt` is imho awful to read.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: right -> left?

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: The last line here is unnecessary imho.

**Review by abcdefg30 — APPROVED** [review]
Looks good otherwise. 👍 Great work!

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (forcecore) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: I don't think people want to see the lengthy vector calculation in the creator function. Can be merged though

Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: I was told on IRC that we rather want it as separate function. However, it'd still be nice if you could rename it to e.g. `calculateVectors` (following the naming conventions).

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: This still needs to be resolved (`Thichkess` -> `Thickness`).

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: I think you can remove `(Tiberian Sun Railgun)`.

**Review by abcdefg30 — COMMENTED** [review]
I think this is good to go once the last two nits are resolved and the commits are squashed.

**Comment by forcecore** [comment]
Rebased, squashed and the last two issues addressed.

**Comment by forcecore** [comment]
* HelixAlphaDeltaPerTick wasn't in effect.
* Player color laser/helix will now use Helix/Laser color's alpha information, not hard coded 128 as alpha.

File: OpenRA.Mods.Common/Graphics/RailgunRenderable.cs
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/RailgunRenderable.cs]: `(railgun.HelixColor.A + ticks * info.HelixAlphaDeltaPerTick).Clamp(0, 255);` would guard against overflow as well as underflow.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (pchote) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: This and the next two appear to be a world-distances, so they should be defined as `WDist`.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (pchote) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: Angles should be defined as `WAngle`.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (pchote) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: Style nit: declare these as `{ get; private set; }` and the compiler will define the backing fields for you, so you can remove the variables defined just above.

File: OpenRA.Mods.Common/Projectiles/Railgun.cs
Comment (pchote) [inline OpenRA.Mods.Common/Projectiles/Railgun.cs]: IMO this looks far too thick ingame, especially when pixel doubled.  32 looks much nicer as a default.

**Review by pchote — COMMENTED** [review]
This looks quite nice.  I have a few minor requests:

**Comment by forcecore** [comment]
Hopefully the last commit :D?

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0553c56c4c77f5cfaf0555e8ffe288886698c8f0

**Pull request merged**
